# EvolGenomics
Tutorial for the 3rd module *Evolutionary Genomics*  of the curse **Bioinformática y Genómica para la Biodiversidad**.

Module instructors:
- **Dr. Eduardo Eizirik**, Pontifícia Universidade Católica do Rio Grande do Sul (PUCRS), Brasil
- **Dr. Jorge Ramírez**, Universidad Nacional Mayor de San Marcos, Perú
- **Dr. Henrique Vieira Figueiró**, (?)
- **Msc. Fernanda Trindade**, Pontifícia Universidade Católica do Rio Grande do Sul (PUCRS), Brasil

## Day 7: Phylogenomics
Today you are going to learn how to use complete genomes to reconstruct the groups phylogeny, to evaluate genealogical discordances and explore the role of hybridization in group's history. We're going to use as example the data from [Figueiró et al., 2017](https://www.science.org/doi/10.1126/sciadv.1700299).

### Phylogenetic inference based on whole-genome sequences
### Genealogical discordance in genome-wide datasets
### Inference of historical hybridization among lineages

## Day 8: Population Genomics
Today you are going to learn how to perform basic populational analyses using complete genomes at low depth, including calling, diversity and homozygosity estimates, population structure and admixture. We're going to use as example the data from [Pecnerova et al., 2021](https://doi.org/10.1016/j.cub.2021.01.064).

### SNP calling vs. genotype likelihoods
The data we'll be using today are **17 leopard individuals from 6 populations across Africa** (see [metadata](https://github.com/ffertrindade/EvolGenomics/blob/main/data/metadata.csv)). These data had been previouslly mapped against domestic cat reference [felCat9](https://www.ncbi.nlm.nih.gov/assembly/GCF_000181335.3) (and you can check the used workflow [here](https://github.com/ffertrindade/EvolGenomics/tree/main/workflows/fastq2bam)) and we'll be using today the [bam](https://github.com/ffertrindade/EvolGenomics/tree/main/data/leopard_population) files containing only the **chromosome A1**.

### Estimates of genome-wide diversity
### Estimates of population structure and admixture
### Estimates of Runs of Homozygosity

## Day 9: Adaptive evolution and demographic history
Today you are going to learn how to perform analyses of natural selection on complete genomes and how to infer historical demography. We're going to use as example the data from [Figueiró et al., 2017](https://www.science.org/doi/10.1126/sciadv.1700299).

### Investigating adaptive evolution based on genome-wide datasets
### Inference of historical demography: PSMC as an example

